### DISCUSSION TITLE
Custom validation when using Forms Plugin

### DISCUSSION THREAD

**Discussion by VisiaMedia** [swup/swup#816] [Help]
Hello community,

I am currently working on a project where I'm utilizing Swup.js, specifically the Forms plugin. However, I've encountered a challenge regarding the validation of forms before they are submitted.

Typically, I would attach an EventListener to the form's "submit" event, using a .preventDefault() method to initially prevent the form from submitting. This approach allows me to validate the form fields using JavaScript, and then proceed with the submission if the validation is successful.

In this case, however, I'm unable to use this approach since the Swup.js Forms plugin seems to take over the "submit" event. I'm looking for a way to delay the event when a user clicks the submit button of the form. After delaying the event, I would like to validate the form fields with JavaScript. If the fields are correctly filled, the form should then be allowed to submit. Conversely, if the fields are incorrectly filled or incomplete, I want to stop the submit event and prevent the form from being sent.

Any guidance or suggestions on how to integrate form validation with the Swup.js Forms plugin would be greatly appreciated.

Thank you in advance for your help and insights.

Best regards,

**Comment by hirasso** [comment]
Hi @VisiaMedia , great that you are bringing this up! Forms Plugin definitely needs a section about how to solve your (very common) scenario. I haven't used Forms Plugin myself in production, yet, but I guess you could go about it like that: Initially, your form doesn't have the [`formSelector`](https://swup.js.org/plugins/forms-plugin/#formselector):

```html
<form action="/">
<button>Submit</button>
</form>
```
Then, in your JavaScript, you could first validate forms on submit, and if a form is valid, add `[data-swup-form]` to it and re-submit it programmatically:

```js
const swup = new Swup({ plugins: [new SwupFormsPlugin()] });
/** catch all form submit events without the need to re-initialize on page:view */
swup.delegateEvent("form:not([data-swup-form])", "submit", onFormSubmit, {
  capture: true,
});

const validateForm = (form) => true; // add your custom logic here

async function onFormSubmit(event) {
  const { delegateTarget: form, submitter } = event;
  event.preventDefault();
  /** validate the form */
  const isValid = await validateForm(form);
  /** add the form selector and re-submit if the form is valid */
  if (isValid) {
    form.setAttribute("data-swup-form", "");
    requestAnimationFrame(() => submitter.click());
  }
}
```

**EDIT**
Turns out `form.submit()` can't be handled by swup, as it doesn't seem to be canceable. Instead, you can use `submitter.click()` to re-submit the form. Just tested this approach, works fine for me (even if I press <kbd>Enter</kbd>). Code updated accordingly.

**Reply by daun** [reply]
There's [form.requestSubmit()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/requestSubmit) which will only submit the form if all validation rules pass. That should be equivalent to doing `submitter.click()`. It won't handle any custom validation logic, however.

**Comment by hirasso** [comment]
@daun any better ideas? I know you have used the plugin in production 🙂

**Comment by hirasso** [comment]
I think introducing an API method for submitting forms through swup would be very helpful.

**Comment by daun** [comment]
Currently, the Forms Plugin isn't built to support custom validation flows. It was created for very basic use cases like search inputs and contact forms that can be done using native browser input validation. Now that it's matured a bit and supports different encoding types and file uploads, we could and should think about supporting custom validation flows.

@VisiaMedia What's your ideal usage scenario here? What would the ideal API look like?

I can image a solution like @hirasso suggested: allowing manual form submissions via swup. That way you can add your own event handlers, perform your own validation, and then manually trigger a form submission. Something like `swup.submit(formEl)`.

**Reply by VisiaMedia** [reply]
Thank you for the prompt response and for considering this functionality in Swup.js. I truly appreciate the community's suggestions and ideas.

My ideal usage scenario is quite straightforward. I don't need to do anything advanced with the API; all I require is the ability to manually take over the submission of a form. If I can validate the form fields using my own JavaScript code and then manually submit the form with something like `swup.submit(formEl)`, that would be perfect. This would provide me the flexibility I need without disrupting the core functionality of the Forms Plugin.

Additionally, I want to express how pleased I am with all the other responses and suggestions. I plan to experiment with the proposed fixes and see how they work out in my project.

Thanks again for your assistance and for considering this addition to Swup.js.
